File: clients/core/src/account.ts
import { b64encode } from './b64';
import { Identity, encodeKey } from './secure';
import { utf8Encode } from './bytes';
import type { Credentials } from './client';

export interface AccountDevice { pub: string; account: string; name: string; role: 'host' | 'phone'; added_at: number; online: boolean }
export interface AccountSession { server: string; token: string; username: string; pub: string; device_seed: string }
export function accountOrigin(raw: string): string {
 const u = new URL(raw.trim());
 if (u.username || u.password || u.search || u.hash || (u.pathname !== '/' && u.pathname !== '') ||
   (u.protocol !== 'https:' && !(u.protocol === 'http:' && ['localhost','127.0.0.1','[::1]'].includes(u.hostname)))) throw new Error('请使用 HTTPS 服务端地址（本机开发可使用 localhost）');
 return u.origin;
}
export class AccountRequestError extends Error {
 constructor(message: string, readonly status: number) { super(message); this.name = "AccountRequestError"; }
}
export async function accountRequest<T>(server: string, token: string, method: string, path: string, data?: unknown): Promise<T> {
 const response = await fetch(accountOrigin(server) + '/v1/account' + path, {
  method, headers: { 'Content-Type': 'application/json', ...(token ? { Authorization: 'Bearer ' + token } : {}) },
  body: data === undefined ? undefined : JSON.stringify(data), signal: AbortSignal.timeout(15_000), redirect: 'error', credentials: 'omit',
 });
 const result = await response.json() as { error?: string };
 if (!response.ok) throw new AccountRequestError(result.error || `账号服务错误 (${response.status})`, response.status);
 return result as T;
}
/** Pass a persisted identity to keep the same device across account logins. */
export async function loginAccount(server: string, username: string, password: string, name: string, register = false, id = Identity.generate()): Promise<{ session: AccountSession; recovery?: string }> {
 username = username.trim().toLowerCase(); server = accountOrigin(server);
 const result = await accountRequest<{ token: string; username: string; pub: string; recovery?: string }>(server, '', 'POST', register ? '/register' : '/login', {
  username, password, name, role: 'phone', pub: encodeKey(id.public_()),
  proof: b64encode(id.signRelayAuth(utf8Encode('wuu/account/enroll/v1:' + username), 'phone')),
 });
 return { session: { server, token: result.token, username: result.username, pub: result.pub, device_seed: b64encode(id.seed()) }, recovery: result.recovery };
}
export function accountCredentials(session: AccountSession, host: AccountDevice): Credentials {
 if (host.role !== 'host' || host.account !== session.username) throw new Error('该电脑不属于当前账号');
 return { v: 1, account_username: session.username, device_seed: session.device_seed, device_name: 'Wuu 手机', host_pub: host.pub, host_name: host.name,
  relay_url: session.server.replace(/^http/, 'ws') + '/v1/connect' };
}

export interface GitHubPending { server: string; request_id: string; verifier: string; oauth_url: string; expires: number; name?: string }
export async function startGitHubLogin(server: string, native = false): Promise<GitHubPending> {
 server = accountOrigin(server);
 if (!globalThis.crypto?.subtle) throw new Error('GitHub 登录需要 HTTPS 网页或 Wuu App');
 const bytes = crypto.getRandomValues(new Uint8Array(32));
 const verifier = b64encode(bytes);
 const hash = new Uint8Array(await crypto.subtle.digest('SHA-256', new TextEncoder().encode(verifier)));
 const result = await accountRequest<{ request_id: string; authorize_url: string; expires_in: number }>(server, '', 'POST', '/github/start', { challenge: b64encode(hash), native });
 const url = new URL(result.authorize_url);
 if (url.origin !== server || url.username || url.password || url.pathname !== '/v1/account/github/authorize' || typeof result.request_id !== 'string' || !/^[A-Za-z0-9_-]{43}$/.test(result.request_id) || url.searchParams.get('state') !== result.request_id) throw new Error('Invalid authorization URL');
 return { server, request_id: result.request_id, verifier, oauth_url: result.authorize_url, expires: Date.now() + 600_000 };
}
export async function pollGitHubLogin(pending: GitHubPending): Promise<{ status: 'pending' | 'authorized'; username?: string }> {
 if (Date.now() >= pending.expires) throw new Error('GitHub login expired; start again');
 return accountRequest(pending.server, '', 'POST', '/github/poll', { request_id: pending.request_id, verifier: pending.verifier });
}
export async function completeGitHubLogin(pending: GitHubPending, username: string, name: string, id: Identity): Promise<AccountSession> {
 const pub = encodeKey(id.public_());
 const result = await accountRequest<{ token: string; username: string; pub: string }>(pending.server, '', 'POST', '/github/complete', {
  request_id: pending.request_id, verifier: pending.verifier, pub, role: 'phone', name,
  proof: b64encode(id.signRelayAuth(utf8Encode('wuu/account/enroll/v1:' + username), 'phone')),
 });
 if (result.username !== username || result.pub !== pub) throw new Error('Invalid enrollment response');
 return { ...result, server: pending.server, device_seed: b64encode(id.seed()) };
}

File: clients/native/android/app/src/main/java/ai/wuu/nativeapp/Account.kt
package ai.wuu.nativeapp

import android.content.Context
import android.security.keystore.KeyGenParameterSpec
import android.security.keystore.KeyProperties
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.suspendCancellableCoroutine
import kotlinx.coroutines.withContext
import okhttp3.Call
import okhttp3.Callback
import okhttp3.HttpUrl
import okhttp3.HttpUrl.Companion.toHttpUrl
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import okhttp3.Response
import org.json.JSONObject
import java.io.IOException
import java.security.KeyStore
import java.util.concurrent.TimeUnit
import javax.crypto.Cipher
import javax.crypto.KeyGenerator
import javax.crypto.SecretKey
import javax.crypto.spec.GCMParameterSpec
import kotlin.coroutines.resume
import kotlin.coroutines.resumeWithException

data class AccountSession(val server: String, val token: String, val username: String, val pub: String, val seed: String, val recovery: String = "") {
    fun encode() = json("server" to server, "token" to token, "username" to username, "pub" to pub, "seed" to seed, "recovery" to recovery)
    companion object {
        fun decode(j: JSONObject) = AccountSession(j.getString("server"), j.getString("token"), j.getString("username"), j.getString("pub"), j.getString("seed"), j.optString("recovery"))
    }
}
class HttpFailure(val status: Int, message: String) : IOException("HTTP $status: $message")

/** Only ciphertext goes into preferences. Keys cannot leave Android Keystore; backups are disabled. */
class Vault(context: Context) {
    private val prefs = context.getSharedPreferences("wuu-private", Context.MODE_PRIVATE)
    private val key: SecretKey = KeyStore.getInstance("AndroidKeyStore").run {
        load(null)
        (getKey("wuu.native.credentials", null) as? SecretKey) ?: KeyGenerator.getInstance(KeyProperties.KEY_ALGORITHM_AES, "AndroidKeyStore").run {
            init(KeyGenParameterSpec.Builder("wuu.native.credentials", KeyProperties.PURPOSE_ENCRYPT or KeyProperties.PURPOSE_DECRYPT)
                .setBlockModes(KeyProperties.BLOCK_MODE_GCM).setEncryptionPaddings(KeyProperties.ENCRYPTION_PADDING_NONE).build())
            generateKey()
        }
    }
    fun read(name: String): JSONObject? {
        val data = prefs.getString(name, null)?.unb64() ?: return null
        require(data.size >= 28)
        val plain = Cipher.getInstance("AES/GCM/NoPadding").run {
            init(Cipher.DECRYPT_MODE, key, GCMParameterSpec(128, data.copyOfRange(0, 12)))
            updateAAD(name.toByteArray()); doFinal(data.copyOfRange(12, data.size))
        }
        return JSONObject(String(plain, Charsets.UTF_8))
    }
    fun write(name: String, value: JSONObject) {
        val encrypted = Cipher.getInstance("AES/GCM/NoPadding").run {
            init(Cipher.ENCRYPT_MODE, key); updateAAD(name.toByteArray()); iv + doFinal(value.toString().toByteArray())
        }
        check(prefs.edit().putString(name, encrypted.b64()).commit()) { "Could not save credentials" }
    }
    fun delete(name: String) { check(prefs.edit().remove(name).commit()) }
    fun identity(server: String, user: String): Identity {
        val name = "identity-" + sha256("$server\n$user".toByteArray()).b64()
        return read(name)?.let { Identity(it.getString("seed").unb64()) } ?: Identity().also {
            write(name, json("seed" to it.seed.b64()))
        }
    }
}

class AccountAPI(server: String) {
    val origin: HttpUrl
    init {
        val raw = java.net.URI(server.trim())
        val parsed = server.trim().toHttpUrl()
        val loopback = parsed.host in setOf("localhost", "127.0.0.1", "::1")
        require((parsed.isHttps || loopback) && raw.rawUserInfo == null &&
            (raw.rawPath.isNullOrEmpty() || raw.rawPath == "/") && raw.rawQuery == null && raw.rawFragment == null) { "请输入 HTTPS 服务器地址，不含路径" }
        origin = parsed
    }
    companion object {
        val client = OkHttpClient.Builder().followRedirects(false).followSslRedirects(false)
            .connectTimeout(15, TimeUnit.SECONDS).readTimeout(30, TimeUnit.SECONDS).callTimeout(30, TimeUnit.SECONDS).build()
    }
    suspend fun request(path: String, token: String? = null, body: JSONObject? = null, query: Map<String, String> = emptyMap(), method: String? = null): JSONObject {
        val url = origin.newBuilder().encodedPath("/v1/account$path").apply { query.forEach { (key, value) -> addQueryParameter(key, value) } }.build()
        val req = Request.Builder().url(url).apply {
            if (token != null) header("Authorization", "Bearer $token")
            if (body != null) post(body.toString().toRequestBody("application/json".toMediaType()))
            if (method == "DELETE") delete()
        }.build()
        val call = client.newCall(req)
        val response = suspendCancellableCoroutine<Response> { continuation ->
            continuation.invokeOnCancellation { call.cancel() }
            call.enqueue(object : Callback {
                override fun onFailure(call: Call, e: IOException) { if (continuation.isActive) continuation.resumeWithException(e) }
                override fun onResponse(call: Call, response: Response) {
                    continuation.resume(response) { _, value, _ -> value.close() }
                }
            })
        }
        return response.use {
            withContext(Dispatchers.IO) {
                val source = it.body?.source() ?: throw IOException("Empty response")
                require(!source.request(16L * 1024 * 1024 + 1)) { "Response too large" }
                val text = source.readUtf8()
                if (!it.isSuccessful) throw HttpFailure(it.code, runCatching { JSONObject(text).optString("error") }.getOrDefault("Request failed"))
                if (text.isBlank()) JSONObject() else JSONObject(text)
            }
        }
    }
    suspend fun login(username: String, password: String, identity: Identity, name: String, register: Boolean = false): AccountSession {
        val result = request(if (register) "/register" else "/login", body = enrollment(username, identity, name).put("password", password))
        return session(result, username, identity)
    }
    private fun enrollment(user: String, identity: Identity, name: String) = json("username" to user, "name" to name,
        "role" to "phone", "pub" to identity.pub.b64(), "proof" to identity.enrollment(user))
    private fun session(result: JSONObject, user: String, identity: Identity): AccountSession {
        require(result.getString("pub") == identity.pub.b64() && result.getString("username") == user) { "Account identity mismatch" }
        return AccountSession(origin.toString(), result.getString("token"), user, identity.pub.b64(), identity.seed.b64(), result.optString("recovery"))
    }
    suspend fun startGitHub(): JSONObject {
        val verifier = randomBytes(32).b64()
        val result = request("/github/start", body = json("challenge" to sha256(verifier.toByteArray()).b64(), "native" to true))
        val url = result.getString("authorize_url").toHttpUrl()
        require(url.scheme == origin.scheme && url.host == origin.host && url.port == origin.port && url.username.isEmpty() && url.password.isEmpty() &&
            url.encodedPath == "/v1/account/github/authorize" && url.queryParameter("state") == result.getString("request_id")) { "Invalid authorization URL" }
        return result.put("server", origin.toString()).put("verifier", verifier)
            .put("expires", System.currentTimeMillis() + result.getLong("expires_in") * 1000)
    }
    suspend fun pollGitHub(pending: JSONObject) = request("/github/poll", body = json("request_id" to pending.getString("request_id"), "verifier" to pending.getString("verifier")))
    suspend fun completeGitHub(pending: JSONObject, user: String, identity: Identity, name: String): AccountSession {
        val result = request("/github/complete", body = enrollment(user, identity, name)
            .put("request_id", pending.getString("request_id")).put("verifier", pending.getString("verifier")))
        return session(result, user, identity)
    }
}
